Source/Randomizer.cpp

```diff
@@ -132,16 +142,75 @@ void Randomizer::Randomize() {
-    Randomize(upDownPanelsSetThree, SWAP::LINES | SWAP::COLORS);
-    Randomize(upDownPanelsSetFour, SWAP::LINES | SWAP::COLORS);
-    Randomize(leftForwardRightPanelsSetOne, SWAP::LINES | SWAP::COLORS);
-    Randomize(leftForwardRightPanelsSetTwo, SWAP::LINES | SWAP::COLORS);
-
-    Randomize(quarryLaserOptions, SWAP::LINES | SWAP::COLORS);
-    Randomize(squarePanels, SWAP::LINES | SWAP::COLORS);
+    if (_doubleRandomizer) {
+        // The pool that Tutorial Back Left is in has some panels that may not
+        // be solveable in the down position after using Sigma's randomizer. We
+        // do not want to swap any such panel with Tutorial Back Left, because
+        // that would make it impossible to exit Tutorial.
+        const std::set<int> bannedTutorialBackLeftPanels = {
+            0x00070, // Symmetry Island Fading Lines 5
+            0x01E5A, // Mill Entry Door Left
+            0x00072, // Symmetry Island Fading 3
+            0x00076, // Symmetry Island Fading 7
+            0x3C125, // Mill Control Room Extra Panel
+            0x09E85, // Tunnels Into Town Big Triangles
+        };
+        std::vector<int> upDownPanelsSetThreeDoubleMode = copyWithoutElements(upDownPanelsSetThree, bannedTutorialBackLeftPanels);
+        Randomize(upDownPanelsSetThreeDoubleMode, SWAP::LINES | SWAP::COLORS);
+        // The four pivot panels in Treehouse must be solveable in the up, left,
+        // and right positions. However, the other panels in the pools those
+        // panels are found in may not be solveable in all three directions
+        // after using Sigma's randomizer. For safety, we will only swap the
+        // pivot panels amongst themselves.
+        Randomize(treehousePivotSet, SWAP::LINES | SWAP::COLORS);
+        // This will additionally shuffle the UTM perspective puzzles amongst
+        // themselves.
+        Randomize(utmPerspectiveSet, SWAP::LINES | SWAP::COLORS);
+        // In order to prevent a situation where access to a Symmetry Laser
+        // Yellow is blocked by its corresponding Blue panel, we will only
+        // shuffle these six panels amongst themselves.
+        Randomize(symmetryLaserYellows, SWAP::LINES | SWAP::COLORS);
+        Randomize(symmetryLaserBlues, SWAP::LINES | SWAP::COLORS);
+        // Prevent the fourth UTM perspective puzzle from being shuffled, as it
+        // is unknown whether any of the other puzzles in these pools can be
+        // solved there after using Sigma's randomizer.
+        const std::set<int> bannedUpDownPanels {
+            0x288AA, // UTM Perspective 4
+        };
+        std::vector<int> upDownPanelsSetZeroDoubleMode = copyWithoutElements(upDownPanelsSetZero, bannedUpDownPanels);
+        std::vector<int> upDownPanelsSetOneDoubleMode = copyWithoutElements(upDownPanelsSetOne, bannedUpDownPanels);
+        std::vector<int> upDownPanelsSetTwoDoubleMode = copyWithoutElements(upDownPanelsSetTwo, bannedUpDownPanels);
```

Comment:
Even though you're re-using this variable, I don't think it really helps readability here to have a separate set. I would just add a comment and then focus on clean code.
```suggestion
        // 0x288AA UTM Perspective 4
        std::vector<int> upDownPanelsSetZeroDoubleMode = copyWithoutElements(upDownPanelsSetZero, {0x288AA});
        std::vector<int> upDownPanelsSetOneDoubleMode = copyWithoutElements(upDownPanelsSetOne, {0x288AA});
        std::vector<int> upDownPanelsSetTwoDoubleMode = copyWithoutElements(upDownPanelsSetTwo, {0x288AA});
```